src/Microsoft.ML.Data/Scorers/FeatureContributionCalculationTransform.cs

```diff
@@ -0,0 +1,500 @@
+                ctx.Writer.WriteBoolByte(_normalize);
+                ctx.Writer.WriteBoolByte(Stringify);
+            }
+
+            public Delegate GetTextContributionGetter(IRow input, int colSrc, VBuffer<ReadOnlyMemory<char>> slotNames)
+            {
+                Contracts.CheckValue(input, nameof(input));
+                Contracts.Check(0 <= colSrc && colSrc < input.Schema.ColumnCount);
+                var typeSrc = input.Schema.GetColumnType(colSrc);
+
+                Func<IRow, int, VBuffer<ReadOnlyMemory<char>>, ValueGetter<ReadOnlyMemory<char>>> del = GetTextValueGetter<int>;
+                var meth = del.GetMethodInfo().GetGenericMethodDefinition().MakeGenericMethod(typeSrc.RawType);
+                return (Delegate)meth.Invoke(this, new object[] { input, colSrc, slotNames });
+            }
+
+            public Delegate GetContributionGetter(IRow input, int colSrc)
+            {
+                Contracts.CheckValue(input, nameof(input));
+                Contracts.Check(0 <= colSrc && colSrc < input.Schema.ColumnCount);
+
+                var typeSrc = input.Schema.GetColumnType(colSrc);
+                Func<IRow, int, ValueGetter<VBuffer<float>>> del = GetValueGetter<int>;
+
+                // REVIEW: Assuming Feature contributions will be VBuffer<float>.
+                // For multiclass LR it needs to be(VBuffer<float>[].
+                var meth = del.GetMethodInfo().GetGenericMethodDefinition().MakeGenericMethod(typeSrc.RawType);
+                return (Delegate)meth.Invoke(this, new object[] { input, colSrc });
+            }
+
+            private ReadOnlyMemory<char> GetSlotName(int index, VBuffer<ReadOnlyMemory<char>> slotNames)
+            {
+                var count = slotNames.GetValues().Length;
+                _env.Assert(count > index || count == 0 && slotNames.Length > index);
+                var slotName = slotNames.GetItemOrDefault(index);
+                return slotName.IsEmpty
+                    ? new ReadOnlyMemory<char>($"f{index}".ToCharArray())
+                    : slotName;
+            }
+
+            private ValueGetter<ReadOnlyMemory<char>> GetTextValueGetter<TSrc>(IRow input, int colSrc, VBuffer<ReadOnlyMemory<char>> slotNames)
+            {
+                Contracts.AssertValue(input);
+                Contracts.AssertValue(Predictor);
+
+                var featureGetter = input.GetGetter<TSrc>(colSrc);
+                var map = Predictor.GetFeatureContributionMapper<TSrc, VBuffer<float>>(_topContributionsCount, _bottomContributionsCount, _normalize);
+
+                var features = default(TSrc);
+                var contributions = default(VBuffer<float>);
+                return
+                    (ref ReadOnlyMemory<char> dst) =>
+                    {
+                        featureGetter(ref features);
+                        map(in features, ref contributions);
+                        var indices = new Span<int>();
```

Comment:
This code is pretty wrong.  You are making an empty span, and then trying to copy things into it (which certainly will throw an exception if the thing you are trying to copy has any elements).

Are we testing this code? I don't see how it could be working.

Reply:
Would the [tests](https://github.com/dotnet/machinelearning/pull/1677/files#diff-3b1795192d6c94f14425c0694f349d42) cover this part of the code?